exporter/exporterhelper/batch_sender.go

```diff
@@ -0,0 +1,461 @@
+// Do not mutate the requests passed to the function if error can be returned after mutation.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type BatchMergeFunc func(context.Context, Request, Request) (Request, error)
+
+// BatchMergeSplitFunc is a function that merge and/or splits a request into multiple requests based on the provided
+// limit of maximum number of items. All the returned requests MUST have a number of items that does not exceed the
+// maximum number of items. Size of the last returned request MUST be less or equal than the size of any other returned
+// request. The original request MUST not be mutated if error is returned. The length of the returned slice MUST not
+// be 0. The optionalReq argument can be nil, make sure to check it before using. maxItems argument is guaranteed to be
+// greater than 0. Context will be propagated from the original request.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type BatchMergeSplitFunc func(ctx context.Context, optionalReq Request, req Request, maxItems int) ([]Request, error)
+
+// IdentifyBatchFunc returns an identifier for a Request batch. This function can be used to separate particular
+// Requests into different batches. Batches with different identifiers will not be merged together.
+// Provided context can be used to extract information from the context and use it as a part of the identifier as well.
+// This function is optional. If not provided, all Requests will be batched together.
+// This API is at the early stage of development and may change without backward compatibility
+// until https://github.com/open-telemetry/opentelemetry-collector/issues/8122 is resolved.
+type IdentifyBatchFunc func(ctx context.Context, r Request) string
+
+type shardGetter interface {
+	shard(*request) (*shard, error)
+}
+
+type Batcher struct {
+	cfg            MergeSplitBatcherConfig
+	mergeFunc      BatchMergeFunc
+	mergeSplitFunc BatchMergeSplitFunc
+
+	shardGetter
+
+	shutdownCh chan struct{}
+	goroutines sync.WaitGroup
+	logger     *zap.Logger
+
+	export func(*request) error
+}
+
+// shutdown is invoked during service shutdown.
+func (b *Batcher) shutdown() {
+	close(b.shutdownCh)
+
+	// Wait until all goroutines are done.
+	b.goroutines.Wait()
+}
+
+// newShard gets or creates a batcher corresponding with attrs.
+func (b *Batcher) newShard() *shard {
+	s := &shard{
+		batcher:    b,
+		newRequest: make(chan *request, runtime.NumCPU()),
+	}
+	b.goroutines.Add(1)
+	go s.start()
+	return s
+}
+
+func (b *Batcher) mergeRequests(req1 *request, req2 *request) (*request, error) {
+	r, err := b.mergeFunc(req1.Context(), req1.Request, req2.Request)
```

Comment:
i wonder if it would be valuable to merge req1 and req2 Contexts into a single contex. here behavior may not be deterministic when considering cancelable contexts or contexts with Deadlines

Reply:
@michalpristas, this is a good question. In the current implementation (which is similar to the current batch processor), we should pass only one context from the request that is currently being added to the batch and is blocking the current `Consume` call. We can ignore the context associated with the existing batch because all the other requests are already accumulated in the batch and are not active anymore.

But if the queue is not enabled, we probably want to block all the incoming requests until a batch is ready, as discussed in https://github.com/open-telemetry/opentelemetry-collector/pull/8762. In that case, we probably need some kind of context merging. That option is not implemented yet.

Reply:
I looked into merging contexts of requests that create a batch.

There was a proposal to implement this capability natively in go https://github.com/golang/go/issues/36503, but it was declined. 

We could manually merge the contexts, but it would require spawning another goroutine specifically to handle the context cancellations. I think this is an overkill for this task. I think it's ok just to take the context from the first request, assuming that it has the shortest timeout. I updated the code accordingly. Let me know WDYT.